### PR TITLE
[WIP] Add dynamic import support

### PR DESCRIPTION
**Pull request by arunoda**
With this we are adding [TC39 dynamic import](https://github.com/tc39/proposal-dynamic-import) support to Next.js

## TODO

* [x] Add Unit tests for SameLoopPromise
* [x] Add support for async-renderer
* [x] Add some integration tests
* [x] Add docs

Basically you could any module asynchronously and do anything you want.
You can use utilities like https://github.com/xtuc/async-reactor without any issues

We've also included some utilities to make your life easier.

> You can also try this [example app](https://github.com/arunoda/next.js/tree/90ea471aa7953cce64429e7aa02494ef86e5114e/examples/with-dynamic-import).

#### 1. Create Dynamic Component (Also does SSR)

```js
import dynamic from 'next/dynamic'
const DynamicComponent = dynamic(import('../components/hello'))

export default () => (
  <div>
    <Header />
    <DynamicComponent />
    <p>HOME PAGE is here!</p>
  </div>
)
```

#### 2. With Custom Loading Component

```js
import dynamic from 'next/dynamic'
const DynamicComponentWithCustomLoading = dynamic(
  import('../components/hello2'),
  {
    loading: () => (<p>...</p>)
  }
)

export default () => (
  <div>
    <Header />
    <DynamicComponentWithCustomLoading />
    <p>HOME PAGE is here!</p>
  </div>
)
```

#### 3. No SSR Support

```js
import dynamic from 'next/dynamic'
const DynamicComponentWithNoSSR = dynamic(
  import('../components/hello3'),
  { ssr: false }
)

export default () => (
  <div>
    <Header />
    <DynamicComponentWithNoSSR />
    <p>HOME PAGE is here!</p>
  </div>
)
```

#### 4. With [async-reactor](https://github.com/xtuc/async-reactor) 

> SSR support is not available here

```js
import { asyncReactor } from 'async-reactor'
const DynamicComponentWithAsyncReactor = asyncReactor(async () => {
  const Hello4 = await import('../components/hello4')
  return (<Hello4 />)
})

export default () => (
  <div>
    <Header />
    <DynamicComponentWithAsyncReactor />
    <p>HOME PAGE is here!</p>
  </div>
)
```

### CODE REVIEW

File: lib/dynamic.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import React from 'react'
+
+let currentChunks = []
+
+export default function dynamicComponent (promise, Loading = () => (<p>Loading...</p>)) {
+  return class Comp extends React.Component {
```
Comment (arunoda) [inline lib/dynamic.js]: Try to get a nice `displayName` if possible.

File: lib/dynamic.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    constructor (...args) {
+      super(...args)
+      this.state = { AsyncComponent: null }
+      this.isServer = typeof window === 'undefined'
+      this.loadComponent()
```
Comment (arunoda) [inline lib/dynamic.js]: Implement `ssr=false` support.

File: server/build/babel/plugins/handle-import.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          require.ensure([], (require) => {
+            let m = require(SOURCE)
+            m = m.default || m
+            m.__webpackChunkName = '${args.name}.js'
+            cb(m);
+          }, 'chunks/${args.name}.js');
+        },
+        catch() {}
+      } :
+      {
```
Comment (arunoda) [inline server/build/babel/plugins/handle-import.js]: Try to use a proper `Promise` or proper promise API.

Comment (arunoda) [inline server/build/babel/plugins/handle-import.js]: Also try to handle import() errors.

**Comment by msand** [comment]
Very cool, great work @arunoda

File: lib/dynamic.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import React from 'react'
+
+let currentChunks = []
+
+export default function dynamicComponent (promise, options = {}) {
+  return class Comp extends React.Component {
```
Comment (arunoda) [inline lib/dynamic.js]: Try to add a proper displayName if possible.

**Comment by xtuc** [comment]
I have a working example with dynamic component loading with `async-reactor`. 

I only changed this form your code:

```js
const DynamicComponent = asyncReactor(import('../components/hello1'))

const DynamicComponentWithCustomLoading = asyncReactor(
  import('../components/hello2'),
  () => (<p>...</p>)
)

export default () => (
  <div>
    <DynamicComponent />
    <DynamicComponentWithCustomLoading />
  </div>
)
```

Are you interessted for a PR?

**Comment by arunoda** [comment]
@xtuc thanks. I will add it to this PR's example.

**Comment by xtuc** [comment]
I didn't publish it yet, i'll do it in a moment. I want to change my example a bit, do you mind if  do a PR against your branch?

**Comment by xtuc** [comment]
@arunoda I did a PR on your fork https://github.com/arunoda/next.js/pull/2

**Comment by arunoda** [comment]
@xtuc I've added it to examples.

File: lib/dynamic.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        this.loadComponent()
+      }
+    }
+
+    render () {
+      const { AsyncComponent } = this.state
+      const { LoadingComponent } = this
+      if (!AsyncComponent) return (<LoadingComponent {...this.props} />)
+
+      return <AsyncComponent {...this.props} />
```
Comment (xtuc) [inline lib/dynamic.js]: This is pretty similar to what async-reactor does. Wouldn't you use it? We need to port some of the SSR logic into it.

**Comment by xtuc** [comment]
Great! thanks :+1:

File: lib/dynamic.js (hunk 5 above)
Comment (arunoda) [inline lib/dynamic.js]: We do some Next.js stuff related here for SSR.
We can try to abstract it later on.

**Comment by arunoda** [comment]
We just took this and it's available with Next.js 3.0.
Check this: https://zeit.co/blog/next3-preview

**Pull request closed without merging**